src/peerconnection.cpp

```diff
@@ -53,6 +53,10 @@ PeerConnection::~PeerConnection() {
 	} catch (const std::exception &e) {
 		PLOG_ERROR << e.what();
 	}
+
+	// Force callbacks to run in impl() so that the shared references to it will free
+	// before our dtor ends, otherwise it won't be freed until after this dtor runs.
+	impl()->join();
```

Comment:
I'm afraid joining here may also create a deadlock situation if the user destroys the `PeerConnection` from a callback, either by releasing the last shared_ptr or by calling `rtcDeletePeerConnection` if using the C API.

I think what your are trying to do might not be feasible: you can't wait for the following tasks to finish from one of the tasks as doing so you block the task processor and therefore the following tasks.

Reply:
I think reasonably, no one should be calling delete while inside a callback no?

Close makes more sense, but for this library to be even usable, the user must have a shared_ptr or similar wrapping the original PeerConnection which wouldn't be safe to free in a callback regardless.

Reply:
It's actually common to delete the `PeerConnection` from a callback if the user doesn't have their own event loop, irrelevant of wrapping: typically, the user would have a map of connections by some kind of id from which a connection is removed when its state changes to disconnected or closed. It's also necessary for the C API because calling `rtcDeletePeerConnection` is explicitly allowed in the API reference.